### ISSUE TITLE
reduce API requests by monitoring state using `docker.events`

### ISSUE THREAD

**Issue by minrk** [jupyterhub/dockerspawner#169]
We should follow [kubespawner's example](https://github.com/jupyterhub/kubespawner/pull/63) and monitor container events instead of querying the docker API every time.

Most importantly, if we can avoid `.poll()` making any API requests, instead referring directly to the event stream, that should improve the scalability quite a bit.

**Comment by yuvipanda** [comment]
Yeah, scalability improved massively for kubespawner with this change. +1 for making it happen for docker too!
